File: FWFramework/Classes/Component/Contacts/FWAuthorizeContacts.m
//
//  FWAuthorize.m
//  FWFramework
//
//  Created by wuyong on 17/3/15.
//  Copyright © 2018年 wuyong.site. All rights reserved.
//

#import "FWAuthorizeContacts.h"
#import <Contacts/Contacts.h>
#if FWFrameworkContacts
@import FWFramework;
#else
#import "FWAuthorize.h"
#endif

// iOS9+使用Contacts
@interface FWAuthorizeContacts : NSObject <FWAuthorizeProtocol>

@end

@implementation FWAuthorizeContacts

+ (void)load
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [FWAuthorizeManager registerAuthorize:FWAuthorizeTypeContacts withBlock:^id<FWAuthorizeProtocol>{
            return [[FWAuthorizeContacts alloc] init];
        }];
    });
}

- (FWAuthorizeStatus)authorizeStatus
{
    CNAuthorizationStatus status = [CNContactStore authorizationStatusForEntityType:CNEntityTypeContacts];
    switch (status) {
        case CNAuthorizationStatusRestricted:
            return FWAuthorizeStatusRestricted;
        case CNAuthorizationStatusDenied:
            return FWAuthorizeStatusDenied;
        case CNAuthorizationStatusAuthorized:
            return FWAuthorizeStatusAuthorized;
        case CNAuthorizationStatusNotDetermined:
        default:
            return FWAuthorizeStatusNotDetermined;
    }
}

- (void)authorize:(void (^)(FWAuthorizeStatus status))completion
{
    [[CNContactStore new] requestAccessForEntityType:CNEntityTypeContacts completionHandler:^(BOOL granted, NSError * _Nullable error) {
        FWAuthorizeStatus status = granted ? FWAuthorizeStatusAuthorized : FWAuthorizeStatusDenied;
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(status);
            });
        }
    }];
}

@end

File: FWFramework/Classes/Module/Contacts/FWAuthorizeContacts.swift
//
//  FWAuthorizeContacts.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/27.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import Contacts
#if FWFrameworkSPM
import FWFramework
import FWFrameworkCompatible
#endif

/// 通讯录授权
@objcMembers public class FWAuthorizeContacts: NSObject, FWAuthorizeProtocol, FWAutoloadProtocol {
    public static func autoload() {
        FWAuthorizeManager.registerAuthorize(.contacts) {
            return FWAuthorizeContacts()
        }
    }
    
    public func authorizeStatus() -> FWAuthorizeStatus {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        switch status {
        case .restricted:
            return .restricted
        case .denied:
            return .denied
        case .authorized:
            return .authorized
        default:
            return .notDetermined
        }
    }
    
    public func authorize(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        CNContactStore().requestAccess(for: .contacts) { granted, error in
            let status: FWAuthorizeStatus = granted ? .authorized : .denied
            if completion != nil {
                DispatchQueue.main.async {
                    completion?(status)
                }
            }
        }
    }
}
